### PR TITLE
C Wrapper for IO Trace Plugin

### PR DESCRIPTION
**Pull request by mariuszbarczak**

### CODE REVIEW

File: source/octf/c/iotrace_plugin.cpp [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+        // Poll for all ongoing traces completion
+        while (isTraceOngoing()) {
+            std::this_thread::sleep_for(std::chrono::milliseconds(10));
+        }
+
+        return IOTracePlugin::stopTrace();
+    }
+
+    void push(uint32_t ioQueueId, const void *trace, size_t size) override {
+        if (ioQueueId > getTraceQueueCount()) {
```
Comment (tomaszrybicki) [inline source/octf/c/iotrace_plugin.cpp]: What's the expected use case here? I think we should exit here as arguments are invalid.

File: source/octf/c/iotrace_plugin.cpp [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+        const char *prefix = "[OCTF]";
+        log::cout << log::prefix << prefix;
+        log::debug << log::prefix << prefix;
+        log::verbose << log::prefix << prefix;
+        log::cerr << log::prefix << prefix;
+        log::critical << log::prefix << prefix;
+
+        context->ref_sid = 0;
+        context->tracing_active = false;
+        plugin = new IOTracePluginC(context, cnfg->id, cnfg->io_queue_count);
```
Comment (tomaszrybicki) [inline source/octf/c/iotrace_plugin.cpp]: We check for success of context allocation but here we do not check for success of allocation

File: source/octf/c/iotrace_plugin.cpp [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+}
+
+extern "C" void octf_iotrace_plugin_destroy(
+        octf_iotrace_plugin_context_t *_context) {
+    if (NULL == _context) {
+        return;
+    }
+
+    octf_iotrace_plugin_context_t context = *_context;
+    if (NULL == context) {
```
Comment (tomaszrybicki) [inline source/octf/c/iotrace_plugin.cpp]: I think we can use nullptr here

File: source/octf/trace/IOTracePlugin.cpp [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        , m_tracing(nullptr)
+        , m_ioQueueCount(queueCount) {}
+
+IOTracePlugin::~IOTracePlugin() {}
+
+bool IOTracePlugin::initCustom() {
+    m_tracing =
+            createInterface<InterfaceTraceCreatingImpl>(getNodePath(), this);
+    if (!m_tracing) {
+        throw Exception("Error creating CAS trace plugin interface");
```
[the remark this answers is not in the document]
Comment (mariuszbarczak) [reply source/octf/trace/IOTracePlugin.cpp]: A step which describes starting tracing added.

File: source/octf/c/iotrace_plugin.cpp (hunk 1 above)
Comment (mariuszbarczak) [reply source/octf/c/iotrace_plugin.cpp]: It was intentionally. It applies for cases when someone created the IO trace plugin with one tracing queue, however a program has more then one IO queue. Then the user is able to combine may IO queues into one.

File: doc/examples/c/iotracer/IOTRACER.md [hunk 5]
```diff
@@ -22,24 +23,23 @@ TBD
 
 ## Producing IO traces by C Application
 
-All fun starts with including C IO trace plug-in.
+All fun starts with including the C IO trace plug-in.
```
Comment (tomaszrybicki) [inline doc/examples/c/iotracer/IOTRACER.md]: All the fun

File: doc/examples/c/iotracer/IOTRACER.md [hunk 6]
```diff
@@ -22,24 +23,23 @@ TBD
...
 #include <octf/c/iotrace_plugin.h>
 ~~~
 
-Having that you can power your
-C application by OCTF. But actually in this case you are going to use C wrapper
-for IO tracer plug-in. When linking your program, don't forget about adding
-OCTF library (just add -loctf to linker option).  
+Having that you can power your C application by OCTF. But actually in this case
+you are going to use C wrapper for IO tracer plug-in. When linking your program,
+don't forget about adding OCTF library (just add -loctf to linker option).
```
Comment (tomaszrybicki) [inline doc/examples/c/iotracer/IOTRACER.md]: options

File: doc/examples/c/iotracer/IOTRACER.md [hunk 7]
```diff
@@ -52,12 +52,64 @@ if (octf_iotrace_plugin_create(&cnfg, &plugin)) {
...
+           --stop-trace                          This command will preemptively end collecting of telemetry data from the drive
+~~~
+
+Now you can start tracing by invoking:
+
+~~~{.sh}
+iotrace --c-iotrace-example --start-trace
+~~~
+
+Print help of start tracing command to get more tracing option.
```
Comment (tomaszrybicki) [inline doc/examples/c/iotracer/IOTRACER.md]: options

**Pull request merged**